### ISSUE TITLE
Add more flexibility to `Sigmoid1d`

### ISSUE THREAD

**Issue by mparno** [MeasureTransport/MParT#412]
Currently the `Sigmoid1d` defines basis functions $f_i(x)$ through

$$\begin{aligned}
f_0(x) &= 1\\  
f_1(x) &= x\\ 
f_2(x) &= w_1 g(-b_2(x-c_1))\\
f_3(x) & = w_2 g(b_3(x-c_2)) \\
f_i(x) & = \sum_{j=0}^{i-4} w_{ij} s(x-c_{ij}) \quad \text{ for } i\geq 4
\end{aligned}$$

where $\mathbf{w}$ and $\mathbf{c}$ are weights and centers defined apriori and fixed during training.   I think we should have some option that makes it easier to construct a simpler basis without using the sum-of-sigmoids when $i\geq 4$ or at least without requiring exactly $i-4$ terms.  One option would be to keep the same interface but add a `bool simple` option to the `Sigmoid1d` constructor that results in the following basis functions:

$$\begin{aligned}
f_0(x) &= 1\\  
f_1(x) &= x\\ 
f_2(x) &= w_1 g(-b_2(x-c_1))\\
f_3(x) & = w_2 g(b_3(x-c_2)) \\
f_i(x) & = s(x-c_{i}) \quad \text{ for } i\geq 4
\end{aligned}$$

Alternative, we could specify the centers and weights separately each term, resulting in a maximally flexible representation:

$$
\begin{aligned}
f_0(x) &= 1\\  
f_1(x) &= x\\ 
f_2(x) &= w_1 g(-b_2(x-c_1))\\
f_3(x) & = w_2 g(b_3(x-c_2)) \\
f_i(x) & = \sum_{j=0}^{k_i} w_{ij} s(x-c_{ij}) \quad \text{ for } i\geq 4
\end{aligned}
$$

The main difference between this representation and the current representation is the variable length $k_i$ for each "order" $i$.

**Comment by dannys4** [comment]
So my main issue with the "simple" version is that it's not a well-posed basis when you start using total order multi-index sets. In some sense, with two dimensions, this would give that [0,27] and [0,4] multi-indices have the same approximation capacity, for example. If you *really* want this, set all centers for $f_i$ to be identical, and you get it with the current functionality. For example, pseudo-julia code:
```julia
num_sigmoids = 5
centers_length = 2 + sum(1:num_sigmoids)
centers = rand(num_sigmoids)
centers_mpart = zeros(centers_length)
g_bound = 3
centers_mpart[1:2] = [-g_bound,g_bound]
mpart_idx = 3
for i in 1:num_sigmoids
     for j in 1:i
        centers_mpart[mpart_idx] = centers[i]
        mpart_idx += 1
     end
end
```

**Comment by dannys4** [comment]
I think that there could be something worthwhile for the more complicated example, but I personally would put this in a v3.x release (or at least, I'd need to see an example where this worked and the original construction doesn't). Frankly, the examples on the RMVE are sparse as-is, and I need to build out where the current implementation works or not.

**Comment by mparno** [comment]
You've probably thought about this a lot as you've implement all this stuff, but I still have some reservations that I'd like to flush out.   Here are the questions on the top of my mind:
1. How does the "complexity" of a single basis function relate to "approximation capacity"?  Without training the weights $w_{ij}$, I'm not sure the order of a single multiindex is ever really interpretable in terms of "approximation capacity".   Regardless of the basis being used, the [0,27] and [0,4] multiindices both refer to a single function and themselves can only approximate functions that are a simple scaling of the basis function.   To me, it's only when we look at a set of multiindices that "approximation capacity" starts to make sense.   With that in mind, I'd rather pull out all the terms currently summed up in the `Sigmoid1d` class and use them as separate terms in the expansion itself.
2. Should we ever be using total order multiindex sets with sigmoidal bases?   I don't think so.  This is a useful way to limit the number of terms in polynomial expansions, but I don't think it has any meaningful interpretation when we start to look at local bases like sigmoids or wavelets.   We would need to use more complicated ["neighborhoods" ](https://github.com/MeasureTransport/MParT/blob/release/MParT/MultiIndices/MultiIndexNeighborhood.h) to correctly define a hierarchy of local bases.   We'd probably want that for ATM or other adaptive techniques, but I don't think it's necessary or important for standard map training where we fix the basis before training.   I think we'd be fine to interpret the "order" 24 sigmoid basis function as simply the sigmoid center at center `c[24]`.

**Comment by dannys4** [comment]
Great points. An idea here then would be more of a multi-indexing funny business where we build a TO mset with respect to the off-diagonal and then create a tensor-product of that with a simple `0:P` on the diagonal (for max degree `P`). What are your thoughts on that?

I don't really like imagining what ATM would look like because forward neighborhoods of elements that are non-constant w.r.t. the diagonal would be a nightmare, and I think any discussion of adaptability for Sigmoids should be off the table because of this for the time being.

**Comment by mparno** [comment]
Your tensor-product idea sounds great to me.  That's definitely more in line with how I've been thinking about the Sigmoid basis.    As for the v3.0 release, I think we need to at least expose this capability via the `CreateSigmoidComponent` factory function.

**Comment by dannys4** [comment]
Would this alleviate the need for the "complex flexible" idea you wrote above with variable $k_i$?

**Comment by mparno** [comment]
Maybe.  Depends on how you define the basis function corresponding to "order" $i$.    I have an idea for a small change to `Sigmoid1d` that will give us the "complex flexible" capability.   I'll throw together a PR so we can see what it would look like and if we want to incorporate it into v3.0.

**Comment by mparno** [comment]
Closed in #414
